x-pack/plugins/enterprise_search/public/applications/workplace_search/views/oauth_authorize/oauth_authorize_logic.ts

```diff
@@ -0,0 +1,157 @@
+    allowOAuthAuthorization: () => null,
+    denyOAuthAuthorization: () => null,
+    setButtonNotLoading: () => null,
+  },
+  reducers: {
+    dataLoading: [
+      true,
+      {
+        setServerProps: () => false,
+      },
+    ],
+    cachedPreAuth: [
+      {} as OAuthPreAuthorization,
+      {
+        setServerProps: (_, serverProps) => transformServerPreAuth(serverProps),
+      },
+    ],
+    buttonLoading: [
+      false,
+      {
+        setButtonNotLoading: () => false,
+        allowOAuthAuthorization: () => true,
+        denyOAuthAuthorization: () => true,
+      },
+    ],
+  },
+  listeners: ({ actions, values }) => ({
+    initializeOAuthPreAuth: async (queryString: string) => {
+      clearFlashMessages();
+      const { http } = HttpLogic.values;
+      const query = parseQueryParams(queryString);
+
+      try {
+        const response = await http.get(route, { query });
+
+        if (response.status === 'redirect') {
+          window.location.replace(response.redirect_uri);
+        } else {
+          actions.setServerProps(response);
+        }
+      } catch (e) {
+        flashAPIErrors(e);
+      }
+    },
+    denyOAuthAuthorization: async () => {
+      const { http } = HttpLogic.values;
+      const { cachedPreAuth } = values;
+
+      try {
+        const response = await http.delete(route, {
+          body: JSON.stringify({
+            client_id: cachedPreAuth.clientId,
+            response_type: cachedPreAuth.responseType,
+            redirect_uri: cachedPreAuth.redirectUri,
+            scope: cachedPreAuth.rawScopes,
+            state: cachedPreAuth.state,
+          }),
+          headers,
+        });
+
+        window.location.replace(response.redirect_uri);
+      } catch (e) {
+        flashAPIErrors(e);
+        actions.setButtonNotLoading();
+      }
+    },
+    allowOAuthAuthorization: async () => {
+      const { http } = HttpLogic.values;
+      const { cachedPreAuth } = values;
+
+      try {
+        const response = await http.post(route, {
+          body: JSON.stringify({
+            client_id: cachedPreAuth.clientId,
+            response_type: cachedPreAuth.responseType,
+            redirect_uri: cachedPreAuth.redirectUri,
+            scope: cachedPreAuth.rawScopes,
+            state: cachedPreAuth.state,
+          }),
+          headers,
+        });
+
+        window.location.replace(response.redirect_uri);
+      } catch (e) {
+        flashAPIErrors(e);
+        actions.setButtonNotLoading();
+      }
+    },
+  }),
+});
+
+const transformServerPreAuth = (serverProps: OAuthPreAuthServerProps): OAuthPreAuthorization => {
+  return {
+    clientId: serverProps.client_id,
+    clientName: serverProps.client_name,
+    redirectUri: serverProps.redirect_uri,
+    responseType: serverProps.response_type,
+    rawScopes: serverProps.scope,
+    scopes: serverProps.scope.split(', '),
+    state: serverProps.state,
+  };
+};
```

Comment:
Optional: You can implicitly return these without having to use `return`

```suggestion
const transformServerPreAuth = (serverProps: OAuthPreAuthServerProps): OAuthPreAuthorization => ({
  clientId: serverProps.client_id,
  clientName: serverProps.client_name,
  redirectUri: serverProps.redirect_uri,
  responseType: serverProps.response_type,
  rawScopes: serverProps.scope,
  scopes: serverProps.scope.split(', '),
  state: serverProps.state,
});
```